src/EditorFeatures/Core/Implementation/IntelliSense/Completion/AsyncCompletionService.cs

```diff
@@ -116,9 +122,38 @@ private bool UseLegacyCompletion(ITextView textView, ITextBuffer subjectBuffer)
                 }
             }
 
-            // Check whether the feature flag (async completion API) is set or this feature is off.
-            if (_newCompletionAPIEnabled == true || !subjectBuffer.GetFeatureOnOffOption(InternalFeatureOnOffOptions.CompletionSet))
+            // Check whether this feature is off.
+            if (!subjectBuffer.GetFeatureOnOffOption(InternalFeatureOnOffOptions.CompletionSet))
+            {
+                return false;
+            }
+
+            // Check whether the feature flag (async completion API) is set.
+            if (_newCompletionAPIEnabled == true)
             {
+                var featureService = _featureServiceFactory.GetOrCreate(textView);
+                if (ShouldEnableCompletion(textView))
+                {
+                    if (!featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+                    {
+                        if (!textView.Properties.TryGetProperty(CompletionDisabledToken, out IFeatureDisableToken completionDisabledToken))
+                        {
+                            throw new InvalidOperationException("Unable to find completionDisabledToken to enable completion for a textView");
+                        }
+
+                        completionDisabledToken.Dispose();
+                        textView.Properties.RemoveProperty(CompletionDisabledToken);
+                    }
+                }
+                else
+                {
+                    if (featureService.IsEnabled(PredefinedEditorFeatureNames.Completion))
+                    {
+                        var completionDisabledToken = featureService.Disable(PredefinedEditorFeatureNames.Completion, this);
```

Comment:
Disabling of feature should happen in `AbstractLanguageService`2.IVsImmediateStatementCompletion2.cs`